### ISSUE TITLE
Add Sonos Move Battery Level to SoCo API

### ISSUE THREAD

**Issue by KurtSanders** [SoCo/SoCo#723]
Add SoCo API functionality to obtain Sonos Move battery level.

![Sonos Move Battery Level](https://user-images.githubusercontent.com/12133562/83397743-1fc6ce80-a3cc-11ea-8aad-f499115328e1.jpg)

**Comment by jjlawren** [comment]
A charging indicator would also be useful.

**Comment by KennethNielsen** [comment]
Looks like a great idea. Is either of you interested in investigating whether it is possible and doing the implementation?

**Comment by jjlawren** [comment]
I would, but I’m starting from scratch on my knowledge of the protocol. Any suggestions on where to start looking for the battery/charger payloads?

**Comment by pwt** [comment]
I used Wireshark to take a superficial look at the interactions between a Sonos controller app and a Move, but the battery level and charging state didn't jump out at me.

**Comment by jjlawren** [comment]
I've looked through the various XML endpoints and nothing seemed to indicate a battery level or charging indicator. Is it possible it's only provided via events?

**Comment by pwt** [comment]
One place battery info can be found is in the speaker support page. Go to:
`http://192.168.0.41:1400/support/review` (substituting the correct IP address for your Move), then click on the name of the Move speaker, then `Battery Status`:

![screenshot_0838](https://user-images.githubusercontent.com/1089749/84040487-9babb280-a99a-11ea-8f5a-6861424e3f56.png)

Note that my Move is running the just released **S2** Sonos software, so it's possibly different under S1.

(The good news is that my preliminary testing indicates that SoCo works with S2.)

**Comment by tksunw** [comment]
I have an S1 Move, and it does not have the "Battery Status" link in the 'support/review' page.

**Comment by KennethNielsen** [comment]
@tksunw you might try and have a look at the address. It might very well have the status page even if the app doesn't expose it.

**Comment by KennethNielsen** [comment]
@pwt can you try and have a look if it has an xml version as well in some variant of xml right after the port and as an extension, like the device page: http://192.168.1.100:1400/xml/device_description.xml

**Comment by pwt** [comment]
> @pwt can you try and have a look if it has an xml version as well in some variant of xml right after the port and as an extension, like the device page: http://192.168.1.100:1400/xml/device_description.xml

Nothing obvious in the device desc. XML. There is, however, a chunk of XML in the page source of the support/review URL I pointed to above:

```
...
<LocalBatteryStatus>
<Data name="Health">GREEN</Data>
<Data name="Level">100</Data>
<Data name="Temperature">NORMAL</Data>
<Data name="PowerSource">SONOS_CHARGING_RING</Data>
</LocalBatteryStatus>
...
```

A reminder that this is with Sonos S2. I only have the one Move, so I can't check what S1 supplies.

**Comment by jjlawren** [comment]
That section is _not_ shown in the S1 `support/review` raw XML.

**Comment by tksunw** [comment]
It does have a device_description.xml page.  No LocalBattery info, though.

[move-s1-device_description.txt](https://github.com/SoCo/SoCo/files/4748675/move-s1-device_description.txt)
[move-s1-status-info.txt](https://github.com/SoCo/SoCo/files/4748677/move-s1-status-info.txt)

**Comment by pwt** [comment]
I have a PR (#756) up for this feature. I've only tested with S2. Based on @jjlawren's report above, it would be useful if someone could test against a Move on an S1 system.

**Comment by jjlawren** [comment]
My system is on S2 so I cannot validate on the legacy system. Should the new feature fail gracefully on a S1 Move?

**Comment by pwt** [comment]
> My system is on S2 so I cannot validate on the legacy system. Should the new feature fail gracefully on a S1 Move?

It will raise a `NotSupportedException`.

**Comment by pwt** [comment]
If anyone wants to do a quick-a-dirty test, there's a similar implementation of this feature present in [soco-cli](https://github.com/avantrec/soco-cli) v0.1.42 now ... pending being able to use the native SoCo one.

Usage: ```sonos <name_of_Move> battery```

Example Output:
```
  Health: GREEN
  Level: 100
  Temperature: NORMAL
  PowerSource: SONOS_CHARGING_RING
```

**Comment by pwt** [comment]
Fixed by #756.
